File: legacy-python/agents/planner.py
"""Planner agent — transforms a user prompt into a product specification."""

from __future__ import annotations

import os
from pathlib import Path

from claude_code_sdk import ClaudeCodeOptions

from .client import AgentResult, run_agent
from security import create_permission_handler
from tools import create_harness_tools


_PROMPTS_DIR = Path(__file__).resolve().parent.parent / "prompts"


async def run_planner(
    user_prompt: str,
    config: dict,
    workspace_dir: str,
) -> AgentResult:
    """Run the planner agent to produce product_spec.json, feature_list.json, and init.sh.

    Parameters
    ----------
    user_prompt:
        The raw user request (1-4 sentences).
    config:
        Parsed config.yaml contents.
    workspace_dir:
        Absolute path to the project workspace directory.
    """
    system_prompt = (_PROMPTS_DIR / "planner_prompt.md").read_text(encoding="utf-8")
    agent_cfg = config["agent_limits"]["planner"]

    options = ClaudeCodeOptions(
        model=config["models"]["planner"],
        system_prompt=system_prompt,
        cwd=workspace_dir,
        max_turns=agent_cfg["max_turns"],
        permission_mode=agent_cfg["permission_mode"],
        allowed_tools=agent_cfg["allowed_tools"],
        mcp_servers={"harness": create_harness_tools(workspace_dir)},
        can_use_tool=create_permission_handler(config, workspace_dir),
    )

    # Build the full prompt with context
    stack_options = config.get("stack", {})
    full_prompt = f"""## User Request

{user_prompt}

## Context

- Your workspace is: {workspace_dir}
- Frontend stack: {stack_options.get('frontend', 'nextjs-tailwind')}
- Backend options: {', '.join(stack_options.get('backend_options', ['api-platform', 'fastapi']))}
- Database: {stack_options.get('database', 'sqlite')}

## Instructions

Produce exactly these 3 files in the workspace root:

1. **product_spec.json** — Full product specification with design system, sprints, and stack choice.
2. **feature_list.json** — Flat array of all features with IDs, sprint assignments, and `"passes": false`.
3. **init.sh** — Idempotent bash script to scaffold the project (directories, dependencies, configs).

Start by analyzing the user request, then design the product, then write the files.
"""

    return await run_agent(full_prompt, options)

File: src/agents/planner.ts
/**
 * Agent Planner — transforme un prompt utilisateur en spécification produit.
 *
 * Port fidèle de la V1 Python (`agents/planner.py`).
 */

import { readFileSync } from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import type { Options } from "@anthropic-ai/claude-agent-sdk";

import { createPermissionHandler } from "../security";
import { createHarnessTools } from "../tools";
import type { Config } from "../types";
import { runAgent, type AgentResult } from "./client";

const PROMPTS_DIR = path.resolve(
  path.dirname(fileURLToPath(import.meta.url)),
  "../../prompts",
);

/**
 * Exécute l'agent planner pour produire product_spec.json, feature_list.json et init.sh.
 *
 * @param userPrompt   La requête utilisateur brute (1-4 phrases).
 * @param config       Contenu parsé de config.yaml.
 * @param workspaceDir Chemin absolu vers le répertoire de workspace du projet.
 */
export async function runPlanner(
  userPrompt: string,
  config: Config,
  workspaceDir: string,
): Promise<AgentResult> {
  const systemPrompt = readFileSync(
    path.join(PROMPTS_DIR, "planner_prompt.md"),
    "utf-8",
  );
  const agentCfg = config.agent_limits.planner;

  const options: Options = {
    model: config.models.planner,
    systemPrompt,
    cwd: workspaceDir,
    maxTurns: agentCfg.max_turns,
    permissionMode: agentCfg.permission_mode as Options["permissionMode"],
    allowedTools: agentCfg.allowed_tools,
    mcpServers: { harness: createHarnessTools(workspaceDir) },
    canUseTool: createPermissionHandler(config, workspaceDir),
  };

  // Construit le prompt complet avec le contexte.
  const stackOptions = config.stack;
  const frontend = stackOptions.frontend ?? "nextjs-tailwind";
  const backendOptions = stackOptions.backend_options ?? ["api-platform", "fastapi"];
  const database = stackOptions.database ?? "sqlite";

  const fullPrompt = `## User Request

${userPrompt}

## Context

- Your workspace is: ${workspaceDir}
- Frontend stack: ${frontend}
- Backend options: ${backendOptions.join(", ")}
- Database: ${database}

## Instructions

Produce exactly these 3 files in the workspace root:

1. **product_spec.json** — Full product specification with design system, sprints, and stack choice.
2. **feature_list.json** — Flat array of all features with IDs, sprint assignments, and \`"passes": false\`.
3. **init.sh** — Idempotent bash script to scaffold the project (directories, dependencies, configs).

Start by analyzing the user request, then design the product, then write the files.
`;

  return runAgent(fullPrompt, options);
}
